lib/rucio/transfertool/fts3_plugins.py

```diff
@@ -0,0 +1,121 @@
+# -*- coding: utf-8 -*-
+# Copyright European Organization for Nuclear Research (CERN) since 2012
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import abstractmethod
+from typing import Callable
+import json
+import sys
+
+from rucio.core.plugins import PolicyPackageAlgorithms
+from rucio.common.config import config_get, config_get_int
+
+from rucio.common.exception import InvalidRequest
+
+
+class FTS3Plugin(PolicyPackageAlgorithms):
+    """
+    Generic plugin for fts3
+    Default plugin, appends an empty dict to the job params
+
+    Creates a policy package type named "fts3_plugin"
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.register(self.__class__.__name__, self.hints)
+        self.transfer_limit = config_get_int(
+            "transfers",
+            option="metadata_byte_limit",
+            raise_exception=False,
+            default=4096,
+        )
+
+    @classmethod
+    def register(cls, name: str, func: Callable) -> None:
+        super()._register("fts3_plugins", algorithm_dict={name: func})
+
+    def set_in_hints(self, *hint_kwargs):
+        raise NotImplementedError
+
+    def _verify_in_format(self, hint_dict: dict) -> None:
+
+        try:
+            hints_json = json.dumps(hint_dict)
+            assert sys.getsizeof(hints_json) < self.transfer_limit
+
+        except TypeError:
+            raise InvalidRequest("Request malformed, cannot encode to JSON")
+        except AssertionError:
+            raise InvalidRequest(
+                f"Request too large, decrease to less than {self.transfer_limit}"
+            )
+
+    def hints(self, *hint_kwargs) -> dict:
+        hints = self.set_in_hints(*hint_kwargs)
+        self._verify_in_format(hints)
+        return self.set_in_hints(*hint_kwargs)
+
+
+class FTS3SchedulingHints(FTS3Plugin):
+    """Query the config file to determine the priority of a transfer based on activity"""
+
+    def __init__(self):
+        super().__init__()
+
+        self.default_prority = config_get(
+            "tape_prority", "default_prority", raise_exception=False, default="20"
+        )
+
+    def _query_session_config(self, activity: str) -> dict[str, str]:
+        """Access the prority for an activity set in config
+
+        Args:
+            activity (str): did activity
+
+        Returns:
+            dict: {'prority': 'Activity Transfer Prority'}
+        """
+        prority = config_get(
```

Comment:
If we start calling config_get functions for each transfer, the database session must be passed through. This will require a refactoring of the whole transfertool interface.